source/microbit/modaudio.cpp

```
@@ -0,0 +1,648 @@
+    if (double_pin) {
+        disable_gpiote(1);
+        nrf_gpio_pin_write(pin1, 0);
+    }
+    //NRF_CLOCK->TASKS_HFCLKSTOP = 1;
+}
+
+static int32_t audio_ticker(void);
+
+
+#define AUDIO_BUFFER_MASK (AUDIO_BUFFER_SIZE-1)
+
+static void init_pin(PinName p0) {
+    pin0 = p0;
+    nrf_gpio_pin_write(pin0, 0);
+    audio_gpiote_init(pin0, 0);
+    audio_ppi_init(0, 0);
+    double_pin = false;
+}
+
+static void init_pins(PinName p0, PinName p1) {
+    pin0 = p0;
+    pin1 = p1;
+    nrf_gpio_pin_write(pin0, 0);
+    nrf_gpio_pin_write(pin1, 0);
+    audio_gpiote_init(pin0, 0);
+    audio_gpiote_init(pin1, 1);
+    audio_ppi_init(0, 1);
+    double_pin = true;
+}
+
+
+
+static void audio_data_fetcher(bool lock) {
+    if (audio_source_iter == NULL) {
+        audio_stop();
+        return;
+    }
+    /* WARNING: We are executing in an interrupt handler.
+     * If an exception is raised here then we must hand it to the VM. */
+    mp_obj_t buffer_obj;
+    if (lock)
+        gc_lock();
+    nlr_buf_t nlr;
+    if (nlr_push(&nlr) == 0) {
+        buffer_obj = mp_iternext_allow_raise(audio_source_iter);
+        nlr_pop();
+        if (lock)
+            gc_unlock();
+    } else {
+        if (lock)
+            gc_unlock();
+        if (!mp_obj_is_subclass_fast(MP_OBJ_FROM_PTR(((mp_obj_base_t*)nlr.ret_val)->type),
+            MP_OBJ_FROM_PTR(&mp_type_StopIteration))) {
+            // an exception other than StopIteration, so set it for the VM to raise later
+            //If memory error, add appropriate message.
+            mp_obj_exception_t *ex = (mp_obj_exception_t *)nlr.ret_val;
+            if (mp_obj_get_type(nlr.ret_val) == &mp_type_MemoryError) {
+                ex->args = (mp_obj_tuple_t *)mp_obj_new_tuple(1, NULL);
+                ex->args->items[0] = mp_obj_new_str("Allocation in interrupt handler",
+                                                    strlen("Allocation in interrupt handler"), false);
+            }
```

Similarly to display, we should just print an error here then let the VM raise the MemoryError as usual.